skipchain/skipchain.go

```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
 	}
 }
 
-// PropagateSkipBlock will save a new SkipBlock
-func (s *Service) propagateSkipBlock(msg network.Message) {
-	sbs, ok := msg.(*PropagateSkipBlocks)
+// PropagateGenesisHandler will save a new SkipBlock
+func (s *Service) propagateGenesisHandler(msg network.Message) {
+	pg, ok := msg.(*PropagateGenesis)
 	if !ok {
 		log.Error("Couldn't convert to slice of SkipBlocks")
 		return
 	}
-	for _, sb := range sbs.SkipBlocks {
-		if !s.blockIsFriendly(sb) {
-			log.Lvlf2("%s: block is not friendly: %x", s.ServerIdentity(), sb.Hash)
+
+	s.db.Store(pg.Genesis)
+
+	if !pg.Genesis.ParentBlockID.IsNull() {
+		sb := s.db.GetByID(pg.Genesis.ParentBlockID)
+		if sb == nil {
+			log.Lvl3("Unknown parent block. Conode could be out of sync.")
 			return
 		}
+
+		sb.ChildSL = append(sb.ChildSL, pg.Genesis.Hash)
+		s.db.Store(sb)
 	}
-	_, err := s.db.StoreBlocks(sbs.SkipBlocks)
-	if err != nil {
-		log.Error(err)
+}
+
+// PropagateForwardLinkHandler will update the latest block with
+// the new forward link and the new block when given
+func (s *Service) propagateForwardLinkHandler(msg network.Message) {
+	pfl, ok := msg.(*PropagateForwardLink)
+	if !ok {
+		log.Error("Couldn't convert to slice of ForwardLinks")
+		return
+	}
+
+	// Get the block to update the list of FLs
+	sb := s.db.GetByID(pfl.ForwardLink.From)
+	if sb == nil {
+		// Happen when the roster has changed
+		err := s.SyncChain(pfl.Roster, pfl.ForwardLink.From)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		sb = s.db.GetByID(pfl.ForwardLink.From)
+		if sb == nil {
+			log.Error("Couldn't get the block to attach the forward link")
+			return
+		}
+	}
+
+	// Because of the sync, the forward link could already be added
+	// but we need to catch the case where the height is much above
+	// the current one which means we missed previous propagation
+	if len(sb.ForwardLink) <= pfl.Height {
+		err := sb.AddForwardLink(pfl.ForwardLink, pfl.Height)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+	}
+
+	// Update the forward link list of the block
+	s.db.Store(sb)
```

Comment:
I'm not clear where this block was validated before we accept it. Does that happen in some other layer of the code? Or is forgotten here?

Reply:
As written above, the block should not even be propagated here. Only taken from the cache (populated by `bftForwardLink`) and then stored if the forward link works out.

Reply:
the skipchain service Store function updates existing blocks with the new forward links after verifying them. If the block as been tempered, the hash is different and it won't be found in the db, then it won't be stored because it can be linked.

That could actually be improved by checking if the id is returned. If not the data is corrupted.